src/include/common/object_pool.h

```diff
@@ -38,23 +68,74 @@ class ObjectPool {
    */
   ~ObjectPool() {
     T *result = nullptr;
-    while (reuse_queue_.Dequeue(&result)) alloc_.Delete(result);
+    while (!reuse_queue_.empty()) {
+      result = reuse_queue_.front();
+      alloc_.Delete(result);
+      reuse_queue_.pop();
+    }
   }
 
   /**
    * Returns a piece of memory to hold an object of T.
-   *
+   * @throw NoMoreObjectException if the object pool fails to fetch memory.
    * @return pointer to memory that can hold T
    */
   T *Get() {
+    SpinLatch::ScopedSpinLatch guard(&latch_);
     T *result = nullptr;
-    if (!reuse_queue_.Dequeue(&result))
-      result = alloc_.New();
-    else
+    if (reuse_queue_.empty()) {
+      if (current_size_ < size_limit_) result = alloc_.New();
+      if (result == nullptr) {
```

Comment:
This is convoluted logic. It should suffice to use an else branch here. Alternatively, flatten the if-statement and throw exception either at the end of function or before the if-block to keep cyclomatic complexity down.

Reply:
clang-tidy will complain if you have an else-clause after throw. I'm not sure how to flatten the if-statement

Reply:
Yes, because they want you to do:

```c++
if (...) throw foo;
...
```

Because you will exit the function immediately after the throw(or return), the else is completely useless.

Reply:
I'm not sure what you want me to do here. Do you want me to do

```c++
if (result != nullptr) { 
    current_size_++;
} else {
    // out of memory
    throw NoMoreObjectException(size_limit_);
}
...
```

Is this better?